### ISSUE TITLE
Observe Vector - CLI

### ISSUE THREAD

**Issue by vector-vic** [timberio/vector#3902]
[v1](https://github.com/timberio/vector-product/issues/14) of the "Observe & monitor Vector" series dealt with internal data collection. It laid the foundation for Vector's internal observability, but did not deal with exposing that data except through Vector's own sinks. While this gives users access to the data, it is still far from a good experience. The user must send that data to a configured sink and then manually setup dashboards to visualize the data.

Our end goal is to provide a web UI that will do this automatically, but before we embark on that project (v3) we want to offer a light-weight CLI that will roughly accomplish the same thing. This is necessary for situatuions where an operator is limited to the command line for insight, such as SSH'ing onto the host Vector is running.

Given that this is useful for Vector operators, we believe this is a simpler first step that will help set the stage for a Vector Web UI in v3.

# User Needs

As a user, I need the...

1. ABILITY to VIEW my configured pipeline, so that I know which components are defined.

2. ABILITY to VIEW performance of Vector at a high-level, so that I know how Vector is generally performing without having to inspect each individual component.

3. ABILITY to VIEW individual component performance, including buffers, so that I can understand exactly how an individual component is performing in my pipeline.

4. ABILITY to VIEW host-level resource usages at all levels, so that I can understand how Vector, and it's components, are using my host's resources.

5. ABILITY to SAMPLE events for specific components, so that I can inspect and debug individual pipeline components.

# Out of scope

1. [CLI-based GUIs](https://raw.githubusercontent.com/fdehau/tui-rs/master/assets/demo.gif)) are out of scope for this project. As much as my inner-nerd wants to do this, all of the presentation details are a distraction from the purpose of this project.

2. Full-blown metrics querying is out of scope for this project. We will be doing this in the future, but we want to use this project to learn more about the requirements.

# Future Work

The intent behind the CLI is to set the stage for v3, where we will work on an admin UI for Vector. Ideally, the communication protocol used for the CLI will be shared by the Web UI.

<p></p>

**Comment by vector-vic** [comment]
Link to feature: https://timber.productboard.com/feature-board/planning/features/5163310

**Comment by jszwedko** [comment]
Scope question:

Do we envision that operators might use this in scripting around `vector`? If so, we should aim to:

* Have the output easily machine parsable (or have a flag to do so)
* Consider any change to the output format as a "breaking change"

I think it could be pretty nice to be able to pipe the output through `awk` and other tools.

**Comment by binarylogic** [comment]
I can see the use case for `tap`, do you think users would want that with `top` though? That'll be tough given that the output is designed for humans and rapidly refreshed.

**Comment by jszwedko** [comment]
I think it could be useful for `top` too. Note that the `top` itself has a `-b` flag for "non-interactive" usage.

**Comment by binarylogic** [comment]
Hm. Would it not be easier to use the GraphQL API or the `internal_metrics` source somehow? It seems like a `-b` flag would be duplicative of those in some ways.

**Comment by leebenson** [comment]
I was very much planning on using the GraphQL API for retrieving data, akin to how the UI would. The stdout format, however, can be whatever we want.

My assumpion was `top` would output in some kind of tabular format, but this was going to be a question once #3778 lands.

**Comment by jszwedko** [comment]
@binarylogic The GraphQL API is definitely an option, but I feel like it is harder to work with on the command line. An analogy might be `aws-cli` vs. issuing raw HTTP API commands.

**Comment by leebenson** [comment]
> The GraphQL API is definitely an option, but I feel like it is harder to work with on the command line.

I might be lacking some context here of certain use-cases, but the assumption I've been operating on since the early RFCs on observing internal metrics and topology, was the idea that `vector top` and the UI would share a common API for the retrieval of data:

The main reasons for this being:

1. There's one API to govern one-time and streaming data.
2. It's typed, both in the Typescript UI, and in the Rust client.
3. Both clients would operate against a common protocol.
4. Remote observability is possible -- e.g. `vector top --url http://some-other-host:8686` -- default to localhost:8686 in the absence of switches.

To clarify my earlier post, this has nothing to do with _formatting output_. This is strictly returning data that can be consumed both via `vector`'s Rust client (that I'm working on in #3778) and in vector-ui.

I imagined there would be additional switches that would control the output of `vector top`. By default, this could be human-readable, tabular data. But it could also be CSV, JSON, or other formats.

An example I'd point to is [kubectl's output options](https://gist.github.com/so0k/42313dbb3b547a0f51a547bb968696ba).

My point being -- API protocol != format.

This might be obvious, but I wanted to make sure we weren't confusing or imagining constraints between data retrieval, and how it ultimately looks in the console.

**Comment by binarylogic** [comment]
> was the idea that vector top and the UI would share a common API for the retrieval of data

That's correct.

**Comment by jszwedko** [comment]
> This might be obvious, but I wanted to make sure we weren't confusing or imagining constraints between data retrieval, and how it ultimately looks in the console.

:+1: agreed. Apologies if I muddied the waters a bit. I'm specifically talking about the output format of `vector top` or other CLI commands. I also expected it'd retrieve the data from the running Vector process via the exposed API.

**Comment by binarylogic** [comment]
Nice work on closing this out @leebenson
